decision-records/2025-04-16--redaction-processor-for-otelcol.md

```diff
@@ -0,0 +1,160 @@
+
+ - Remove any attributes not included in a predefined list of permitted attributes.
+ - Strip confidential information from telemetry to prevent accidental data exposure.
+ - Mask or obfuscate sensitive attribute values that match standard patterns.
+
+
+
+### Config examples
+
+#### Log record in plain text format
+
+In a log line like this one:
+
+```
+2025-04-14T10:45:32Z INFO User john.doe@example.com logged in with token abcdef123456
+```
+
+We need to::
+
+ - Mask the email → `***`
+ - Mask the token → `***`
+
+The `redaction` processor config block looks like:
+
+```yaml
+processors:
+  redaction/redact_email_token:
+    rules:
+      - name: redact_email
+        pattern: '[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}'
+        replacement: '***'
+      - name: redact_token
+        pattern: 'token\s+\w+'
+        replacement: 'token ***'
+```
+
+
+#### Log record in JSON format
+
+```json
+{"timestamp": "2025-04-14T10:45:32Z", "level": "INFO", "user": "john.doe@example.com", "token": "abcdef123456", "action": "login"}
+```
+
+We need to::
+
+ - Mask the email → `***`
+ - Mask the token → `***`
+
+
+The `redaction` processor config block looks like:
+
+```yaml
+processors:
+  redaction/redact_email_token:
+    rules:
+      - name: redact_email
+        pattern: '"user":\s*"[^"]+@[^"]+"'
+        replacement: '"user": "***"'
+      - name: redact_token
+        pattern: '"token":\s*"[^"]+"'
+        replacement: '"token": "***"'
+```
+
+
+#### Use these configs in a `pipeline`
+
+In order to use one of these configs for the `redaction` procesor in a pipeline, we need to add it like this:
+
+```yaml
+service:
+  pipelines:
+    logs/0:
+      receivers:
+      - loki
+      processors:
+      - redaction/redact_email_token
+      exporters:
+      - loki/0
+```
+
+### How Loki stores the redacted logs:
+
+Loki stores the modified logs these ways
+
+#### Log record in plain text format:
+```json
+{
+  "stream": {
+    "job": "my-job",
+    "filename": "/var/log/myapp.log",
+    "level": "INFO"
+  },
+  "values": [
+    [
+      "1713081932000000000",
+      "2025-04-14T10:45:32Z INFO User *** logged in with token ***"
+    ]
+  ]
+}
+```
+
+#### Log record in JSON format:
+
+```json
+{
+  "stream": {
+    "job": "my-job",
+    "filename": "/var/log/myapp.log",
+    "level": "INFO"
+  },
+  "values": [
+    [
+      "1713081932000000000",
+      "{\"timestamp\": \"2025-04-14T10:45:32Z\", \"level\": \"INFO\", \"user\": \"***\", \"token\": \"***\", \"action\": \"login\"}"
+    ]
+  ]
+}
+```
+
+## Interaction through `juju` command line interface
+
+The `redaction` processor confis should be defined in a `yaml` file and their mapping to pipelines like this:
+
+```shell
+juju config otelcol redaction_file='@path/to/redaction_processor_config_file.yaml' pipelines_mapping='redaction/redact_email_token=logs/0, redaction/redact_email_token=logs/1`
```

Comment:
I think we should avoid having pipelines mapping:
- We apply the same thing to all logs pipelines.
- For more elaborate setup we'd use the integrator charm approach.

Off-topic impl detail: The config option name should not include `_file`, because technically it doesn't have to come from a file. Maybe `redaction_config`?

Reply:
I have some doubt about avoiding having pipelines mapping.

If I remember correctly, one of the valuable features of redacting logs is the possibility to send raw log to your on premise deployment and redacted logs to public clouds at the same time. So if this is a feature we want, pipeline mapping is needed

Reply:
I suspect it's not going to be easy for an admin to figure out what they need to put in `redaction_mapping`. Seems leaky to expect that level of involvement.
The feature you mention could be impl'd by using integrator charm(s).